akka-persistence-typed/src/main/scala/akka/persistence/typed/javadsl/CommandHandler.scala

```diff
@@ -0,0 +1,91 @@
+/**
+ * Copyright (C) 2018 Lightbend Inc. <https://www.lightbend.com>
+ */
+package akka.persistence.typed.javadsl
+
+import akka.actor.typed.javadsl.ActorContext
+import akka.annotation.InternalApi
+import akka.japi.pf.FI
+import akka.persistence.typed.scaladsl.PersistentBehaviors.EffectImpl
+import akka.util.OptionVal
+
+/**
+ * SAM for reacting on commands
+ *
+ * Used with [[CommandHandlerBuilder]] to setup the behavior of a [[PersistentBehavior]]
+ */
+@FunctionalInterface
+trait CommandHandler[Command, Event, State] {
+  def apply(ctx: ActorContext[Command], state: State, command: Command): Effect[Event, State]
+}
+/**
+ * SAM for reacting on signals
+ *
+ * Used with [[CommandHandlerBuilder]] to setup the behavior of a [[PersistentBehavior]]
+ */
+@FunctionalInterface
+trait CommandToEffect[Command, MsgCommand <: Command, Event, State] {
+  /**
+   * @return An effect for the given command
+   */
+  def apply(ctx: ActorContext[Command], state: State, command: MsgCommand): Effect[Event, State]
+}
+
+object CommandHandlerBuilder {
+  def builder[Command, Event, State](rootCommandClass: Class[Command]): CommandHandlerBuilder[Command, Event, State] =
+    new CommandHandlerBuilder[Command, Event, State](rootCommandClass)
+}
+
+final class CommandHandlerBuilder[Command, Event, State] @InternalApi private[persistence] (rootCommandClass: Class[Command]) {
+
+  private final case class CommandHandlerCase(predicate: Command ⇒ Boolean, handler: CommandToEffect[Command, Command, Event, State])
+
+  private var cases: List[CommandHandlerCase] = Nil
+
+  private def addCase(predicate: Command ⇒ Boolean, handler: CommandToEffect[Command, Command, Event, State]): Unit = {
+    cases = CommandHandlerCase(predicate, handler) :: cases
+  }
+  /**
+   * Match any command which the given `predicate` returns true for
+   */
+  def matchCommand(predicate: FI.TypedPredicate[Command], commandToEffect: CommandToEffect[Command, Command, Event, State]): CommandHandlerBuilder[Command, Event, State] = {
+    addCase(cmd ⇒ predicate.defined(cmd), commandToEffect)
+    this
+  }
+
+  def matchCommand[C <: Command](commandClass: Class[C], commandToEffect: CommandToEffect[Command, C, Event, State]): CommandHandlerBuilder[Command, Event, State] = {
```

Comment:
I think this type bound is also going up in smoke, that's why we have the manual type check in the code instead.

Reply:
This appears to work it ends up being Class<? extends Command>, so not sure we need the check?

Reply:
Maybe there is some difference to the streams API then and I drew the wrong conclusion about using a type parameter of the surrounding class as bound being the problem. Thanks for checking!